server/src/main/java/com/microsoft/java/bs/core/internal/reporter/TestReportReporter.java

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+package com.microsoft.java.bs.core.internal.reporter;
+
+import ch.epfl.scala.bsp4j.TaskId;
+import ch.epfl.scala.bsp4j.TaskStartParams;
+import ch.epfl.scala.bsp4j.TestFinish;
+import ch.epfl.scala.bsp4j.TestStart;
+import ch.epfl.scala.bsp4j.TestStatus;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.OperationResult;
+import org.gradle.tooling.events.ProgressEvent;
+import org.gradle.tooling.events.StartEvent;
+import org.gradle.tooling.events.test.JvmTestOperationDescriptor;
+import org.gradle.tooling.events.test.TestFailureResult;
+import org.gradle.tooling.events.test.TestSkippedResult;
+import org.gradle.tooling.events.test.TestSuccessResult;
+
+import ch.epfl.scala.bsp4j.BuildClient;
+import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
+import ch.epfl.scala.bsp4j.StatusCode;
+import ch.epfl.scala.bsp4j.TaskFinishParams;
+import ch.epfl.scala.bsp4j.TestReport;
+import org.gradle.tooling.internal.consumer.DefaultTestAssertionFailure;
+
+/**
+ * Implements {@link ProgressReporter} to record test results.
+ * Test summary report. e.g. number of fails.
+ */
+public class TestReportReporter extends ProgressReporter {
+
+  private final BuildTargetIdentifier btId;
+  private int successCount;
+  private int skippedCount;
+  private int failureCount;
+  private String exception;
+  private long testDuration;
+
+  /**
+   * initialise.
+   *
+   * @param btId the build target being tested.
+   */
+  public TestReportReporter(BuildTargetIdentifier btId, BuildClient client, String originId) {
+    super(client, originId);
+    this.btId = btId;
+    successCount = 0;
+    skippedCount = 0;
+    failureCount = 0;
+    exception = null;
+    testDuration = 0;
+  }
+
+  @Override
+  public void statusChanged(ProgressEvent event) {
+    if (client != null) {
+      if (event.getDescriptor() instanceof JvmTestOperationDescriptor) {
+        // only report on methods
+        JvmTestOperationDescriptor descriptor = (JvmTestOperationDescriptor) event.getDescriptor();
+        if (descriptor.getClassName() != null && descriptor.getMethodName() != null) {
+          TaskId taskId = getTaskId(descriptor.getDisplayName());
+          if (event instanceof StartEvent) {
+            TaskStartParams startParam = new TaskStartParams(taskId);
+            startParam.setMessage("Start test");
+            startParam.setDataKind("test-start");
+            startParam.setEventTime(event.getEventTime());
+            TestStart testStart = new TestStart(event.getDisplayName());
```

Comment:
Looks like the `event.getDisplayName()` will report event like:

```
Test isBasicTest()(com.example.project.PassingTests) started
```

which contains the class name and method name that the client can be used to target the test item.

What do you think about we constructing the message by ourselves, so that we will not be impacted if the `.getDisplayName()` is changed from upstream?

Reply:
I think BSP isn't perfectly defined here.  It has `Location` which is a position in a file.  The test frameworks that I've seen (`junit`, `sbt/test-interface`) don't provide this info. Instead they provide `className` & `methodName`.
1. We could extends the `TestStart` & `TestFinish` classes to include `className` & `methodName`.
2. We could have a custom `TestFinishDataKind` to hold `className` & `methodName` as there don't appear to be any `TestFinishDataKind` types but that apply only to `TestFinish`, not `TestStart`.
3. We could pass this data encoded as JSON in the `displayName`.
4. We could apply to change the BSP spec.

All seem like fudges except the 4th, but the client does need this info somehow and I'm happy to go down the custom message route.

Reply:
Thank you for the points listed here!

I thought about this question again, extending `TestStart` & `TestFinish` sounds better to me. It's a feasible solution if we want to change BSP spec. On the other hand, an JSON encoded message somehow breaks the semantic meaning of `displayname`.